### ISSUE TITLE
AMDGPU: selectiondag generates broken code when SALU compare result is used by VALU select instruction

### ISSUE THREAD

**Issue by jayfoad** [llvm/llvm-project#59618]
It seems hard to get a small test case for this. This is the best I've come up with so far:
```ll
define void @f(i32 %arg, float* %ptr) {
bb:
  %i = load <2 x i32>, <2 x i32> addrspace(4)* null, align 4294967296
  %i1 = extractelement <2 x i32> %i, i64 1
  %i2 = extractelement <2 x i32> %i, i64 0
  %i3 = lshr i32 %i1, 1
  %i4 = icmp ne i32 %i2, 0
  %i5 = select i1 %i4, i32 %i3, i32 0
  %i6 = sitofp i32 %i5 to float
  %i7 = call float @llvm.fabs.f32(float %i6)
  %i8 = uitofp i1 %i4 to float
  %i9 = lshr i32 1, %i2
  %i10 = select i1 %i4, i32 %i9, i32 0
  %i11 = sitofp i32 %i10 to float
  %i12 = select i1 %i4, i32 %i1, i32 0
  %i13 = sitofp i32 %i12 to float
  br label %bb14

bb14:
  %i15 = phi float [ 0.0, %bb ], [ %i19, %bb14 ]
  %i16 = fadd float %i15, %i8
  %i17 = fadd float %i16, %i7
  %i18 = fadd float %i17, %i11
  %i19 = fadd float %i18, %i13
  %i20 = icmp sgt i32 %arg, 0
  br i1 %i20, label %bb14, label %bb21

bb21:
  store float %i15, float* %ptr, align 4
  ret void
}

declare float @llvm.fabs.f32(float)
```
If I compile this with `llc -march=amdgcn -mcpu=gfx1010 -stop-before=si-fix-sgpr-copies` I get:
```ll
  bb.0.bb:
    successors: %bb.1(0x80000000)
    liveins: $vgpr0, $vgpr1, $vgpr2
  
    %13:vgpr_32 = COPY $vgpr2
    %12:vgpr_32 = COPY $vgpr1
    %11:vgpr_32 = COPY $vgpr0
    %17:sreg_64 = S_MOV_B64 0
    %18:sreg_64_xexec = S_LOAD_DWORDX2_IMM killed %17, 0, 0 :: (invariant load (s64) from `<2 x i32> addrspace(4)* null`, align 4294967296, addrspace 4)
    %19:sreg_32 = COPY %18.sub1
    %20:sreg_32 = COPY %18.sub0
    %15:sreg_32 = S_MOV_B32 0
    S_CMP_LG_U32 %20, %15, implicit-def $scc
    %21:sreg_32 = S_CSELECT_B32 %19, %15, implicit $scc
    %22:sreg_32 = S_MOV_B32 1
    %23:sreg_32 = S_LSHR_B32 %22, %20, implicit-def dead $scc
    S_CMP_LG_U32 %20, %15, implicit-def $scc
    %24:sreg_32 = S_CSELECT_B32 killed %23, %15, implicit $scc
    %25:sreg_32 = S_LSHR_B32 %19, %22, implicit-def dead $scc
    S_CMP_LG_U32 %20, %15, implicit-def $scc
    %26:sreg_64 = REG_SEQUENCE %12, %subreg.sub0, %13, %subreg.sub1
    %14:vreg_64 = COPY %26
    %27:sreg_32 = S_CSELECT_B32 killed %25, %15, implicit $scc
    %28:vgpr_32 = V_CVT_F32_I32_e64 killed %27, 0, 0, implicit $mode, implicit $exec
    %29:sreg_32 = S_MOV_B32 2147483647
    %30:sreg_32 = COPY %28
    %0:sgpr_32 = S_AND_B32 killed %30, killed %29, implicit-def dead $scc ; !!! SCC DEAD HERE !!!
    %31:sgpr_32 = S_MOV_B32 1065353216
    %16:sgpr_32 = S_MOV_B32 0
    %33:sreg_32_xm0_xexec = COPY $scc ; !!! SCC USED HERE !!!
    %34:vgpr_32 = COPY killed %31
    %32:vgpr_32 = V_CNDMASK_B32_e64 0, %16, 0, %34, %33, implicit $exec
    %1:sgpr_32 = COPY %32
    %35:vgpr_32 = V_CVT_F32_UBYTE0_e64 killed %24, 0, 0, implicit $exec
    %2:sgpr_32 = COPY %35
    %36:vgpr_32 = V_CVT_F32_I32_e64 killed %21, 0, 0, implicit $mode, implicit $exec
    %3:sgpr_32 = COPY %36
    %37:sreg_32 = V_CMP_LT_I32_e64 %11, %22, implicit $exec
    %4:vreg_1 = COPY %37
  
  bb.1.bb14:
    successors: %bb.2(0x04000000), %bb.1(0x7c000000)
  
    %5:sreg_32 = PHI %15, %bb.0, %7, %bb.1
    %6:sgpr_32 = PHI %16, %bb.0, %8, %bb.1
    %38:sreg_32 = COPY %4
    %7:sreg_32 = SI_IF_BREAK %38, %5, implicit-def dead $scc
    %39:vgpr_32 = nofpexcept V_ADD_F32_e64 0, %6, 0, %1, 0, 0, implicit $mode, implicit $exec
    %40:vgpr_32 = nofpexcept V_ADD_F32_e64 0, killed %39, 0, %0, 0, 0, implicit $mode, implicit $exec
    %41:vgpr_32 = nofpexcept V_ADD_F32_e64 0, killed %40, 0, %2, 0, 0, implicit $mode, implicit $exec
    %42:vgpr_32 = nofpexcept V_ADD_F32_e64 0, killed %41, 0, %3, 0, 0, implicit $mode, implicit $exec
    %8:sgpr_32 = COPY %42
    SI_LOOP %7, %bb.1, implicit-def dead $exec, implicit-def dead $scc, implicit $exec
    S_BRANCH %bb.2
  
  bb.2.bb21:
    %9:vgpr_32 = PHI %6, %bb.1
    %10:sreg_32 = PHI %7, %bb.1
    SI_END_CF %10, implicit-def dead $exec, implicit-def dead $scc, implicit $exec
    FLAT_STORE_DWORD %14, %9, 0, 0, implicit $exec, implicit $flat_scr :: (store (s32) into %ir.ptr)
    SI_RETURN
```
I've added `!!!` comments to indicate the problem. The problem is that `%32 = V_CNDMASK` wants to use the condition that was calculated by one of the earlier `S_CMP_LG_U32` instructions, so it inserts a copy from `$scc`, but it does it at a point where `$scc` is dead because it has been clobbered by the `%0 = S_AND_B32` instruction.

**Comment by jayfoad** [comment]
My patch to `InstrEmitter::EmitCopyFromReg` in https://reviews.llvm.org/D124852 will fix this problem.

\+@jrbyrnes you worked on https://reviews.llvm.org/D128681

\+@alex-t you worked on https://reviews.llvm.org/D133593

**Comment by jayfoad** [comment]
Fixed by e73b35699be0166d2de8cbb7cb6ac544a3e63f5f
